File: sass-kb-admin/src/services/docService.ts
import api from './api';
import type { PageResult, ApiResponse } from '@/types/common';

export interface Space {
  id: string;
  tenantId: string;
  name: string;
  description: string;
  icon: string;
  type: string;
  sortOrder: number;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface SpaceTreeNode {
  id: string;
  name: string;
  type: 'space' | 'folder' | 'doc';
  status?: string;
  updatedAt?: string;
  children?: SpaceTreeNode[];
}

export interface Folder {
  id: string;
  spaceId: string;
  parentId: string;
  name: string;
  sortOrder: number;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface Document {
  id: string;
  spaceId: string;
  folderId: string;
  tenantId: string;
  title: string;
  contentJson: string;
  contentHtml: string;
  status: string;
  version: number;
  createdBy: string;
  updatedBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface DocSaveRequest {
  title: string;
  contentJson: string;
  contentHtml: string;
  version: number;
  changeSummary?: string;
  updatedBy?: string;
}

export interface DocumentVersion {
  id: string;
  documentId: string;
  versionNumber: number;
  contentJson: string;
  contentHtml: string;
  changeSummary: string;
  createdBy: string;
  createdAt: string;
}

export const spaceApi = {
  list: (keyword?: string) =>
    api.get<any, ApiResponse<Space[]>>('/space/list', { params: { keyword } }),
  create: (data: Partial<Space>) =>
    api.post<any, ApiResponse<Space>>('/space', data),
  update: (id: string, data: Partial<Space>) =>
    api.put<any, ApiResponse<Space>>(`/space/${id}`, data),
  delete: (id: string) =>
    api.delete<any, ApiResponse<void>>(`/space/${id}`),
  getTree: (id: string) =>
    api.get<any, ApiResponse<SpaceTreeNode[]>>(`/space/${id}/tree`),
};

export const folderApi = {
  create: (data: Partial<Folder>) =>
    api.post<any, ApiResponse<Folder>>('/folder', data),
  update: (id: string, data: Partial<Folder>) =>
    api.put<any, ApiResponse<Folder>>(`/folder/${id}`, data),
  delete: (id: string) =>
    api.delete<any, ApiResponse<void>>(`/folder/${id}`),
  move: (id: string, targetParentId: string) =>
    api.put<any, ApiResponse<void>>(`/folder/${id}/move?targetParentId=${targetParentId}`),
  sort: (items: { id: string; sortOrder: number }[]) =>
    api.put<any, ApiResponse<void>>('/folder/sort', items),
};

export const docApi = {
  create: (data: Partial<Document>) =>
    api.post<any, ApiResponse<Document>>('/doc', data),
  getById: (id: string) =>
    api.get<any, ApiResponse<Document>>(`/doc/${id}`),
  save: (id: string, data: DocSaveRequest) =>
    api.put<any, ApiResponse<Document>>(`/doc/${id}`, data),
  delete: (id: string) =>
    api.delete<any, ApiResponse<void>>(`/doc/${id}`),
  updateStatus: (id: string, status: string) =>
    api.put<any, ApiResponse<Document>>(`/doc/${id}/status?status=${status}`),
  getVersions: (id: string) =>
    api.get<any, ApiResponse<DocumentVersion[]>>(`/doc/${id}/versions`),
  getVersion: (id: string, versionNumber: number) =>
    api.get<any, ApiResponse<DocumentVersion>>(`/doc/${id}/versions/${versionNumber}`),
  diff: (id: string, v1: number, v2: number) =>
    api.get<any, ApiResponse<{ v1: string; v2: string; v1Version: string; v2Version: string }>>(`/doc/${id}/diff`, { params: { v1, v2 } }),
  list: (params: { spaceId: string; folderId?: string; page?: number; size?: number; keyword?: string; status?: string }) =>
    api.get<any, ApiResponse<PageResult<Document>>>('/doc/list', { params }),
  restoreVersion: (id: string, versionNumber: number) =>
    api.put<any, ApiResponse<Document>>(`/doc/${id}/versions/${versionNumber}/restore`),
  recent: (limit?: number) =>
    api.get<any, ApiResponse<Document[]>>('/doc/recent', { params: { limit } }),
};

File: sass-kb-server/sass-kb-doc/src/main/java/com/sass/kb/doc/service/DocService.java
package com.sass.kb.doc.service;

import cn.hutool.core.util.IdUtil;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.sass.kb.common.exception.BizException;
import com.sass.kb.common.result.PageResult;
import com.sass.kb.common.util.HtmlSanitizer;
import com.sass.kb.doc.dto.DocSaveRequest;
import com.sass.kb.doc.entity.Document;
import com.sass.kb.doc.entity.DocumentVersion;
import com.sass.kb.doc.mapper.DocumentMapper;
import com.sass.kb.doc.mapper.DocumentVersionMapper;
import com.sass.kb.tenant.context.TenantContext;
import lombok.RequiredArgsConstructor;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class DocService {

    private final DocumentMapper documentMapper;
    private final DocumentVersionMapper documentVersionMapper;
    private final RabbitTemplate rabbitTemplate;

    public Document create(Document doc, String createdBy) {
        String tenantId = TenantContext.getCurrentTenantId();
        doc.setId(IdUtil.fastSimpleUUID());
        doc.setTenantId(tenantId);
        doc.setStatus(doc.getStatus() != null ? doc.getStatus() : "draft");
        doc.setVersion(1);
        doc.setCreatedBy(createdBy);
        doc.setUpdatedBy(createdBy);
        documentMapper.insert(doc);
        return doc;
    }

    public Document getById(String id) {
        Document doc = documentMapper.selectById(id);
        if (doc == null) {
            throw new BizException(404, "文档不存在");
        }
        return doc;
    }

    public Document save(String docId, DocSaveRequest req) {
        Document current = documentMapper.selectById(docId);
        if (current == null) {
            throw new BizException(404, "文档不存在");
        }

        // 乐观锁检查
        if (req.getVersion() == null) {
            throw new BizException(400, "版本号不能为空");
        }
        if (!req.getVersion().equals(current.getVersion())) {
            throw new BizException(409, "版本冲突：文档已被他人修改，请刷新后重试");
        }

        // 保存旧版本
        DocumentVersion dv = new DocumentVersion();
        dv.setId(IdUtil.fastSimpleUUID());
        dv.setDocumentId(docId);
        dv.setVersionNumber(current.getVersion());
        dv.setContentJson(current.getContentJson());
        dv.setContentHtml(current.getContentHtml());
        dv.setChangeSummary(req.getChangeSummary());
        dv.setCreatedBy(req.getUpdatedBy());
        documentVersionMapper.insert(dv);

        // 更新当前版本
        current.setTitle(req.getTitle());
        current.setContentJson(req.getContentJson());
        current.setContentHtml(HtmlSanitizer.sanitize(req.getContentHtml()));
        current.setVersion(current.getVersion() + 1);
        if (req.getUpdatedBy() != null) {
            current.setUpdatedBy(req.getUpdatedBy());
        }
        documentMapper.updateById(current);

        // 异步更新 ES
        rabbitTemplate.convertAndSend("doc.change", "doc.updated", docId);

        return documentMapper.selectById(docId);
    }

    public void delete(String id) {
        Document doc = documentMapper.selectById(id);
        if (doc == null) {
            throw new BizException(404, "文档不存在");
        }
        documentMapper.deleteById(id);
        // 异步更新 ES (删除索引)
        rabbitTemplate.convertAndSend("doc.change", "doc.deleted", id);
    }

    public Document updateStatus(String id, String status) {
        Document doc = documentMapper.selectById(id);
        if (doc == null) {
            throw new BizException(404, "文档不存在");
        }
        // 只更新 status 字段，不触碰 version，避免与其他并发保存冲突
        Document update = new Document();
        update.setId(id);
        update.setStatus(status);
        documentMapper.updateById(update);
        doc.setStatus(status);
        // 状态变更也同步 ES
        rabbitTemplate.convertAndSend("doc.change", "doc.updated", id);
        return doc;
    }

    public List<DocumentVersion> getVersions(String docId) {
        Document doc = documentMapper.selectById(docId);
        if (doc == null) {
            throw new BizException(404, "文档不存在");
        }
        return documentVersionMapper.selectList(
                new LambdaQueryWrapper<DocumentVersion>()
                        .eq(DocumentVersion::getDocumentId, docId)
                        .orderByDesc(DocumentVersion::getVersionNumber));
    }

    public DocumentVersion getVersion(String docId, Integer versionNumber) {
        DocumentVersion dv = documentVersionMapper.selectOne(
                new LambdaQueryWrapper<DocumentVersion>()
                        .eq(DocumentVersion::getDocumentId, docId)
                        .eq(DocumentVersion::getVersionNumber, versionNumber));
        if (dv == null) {
            throw new BizException(404, "版本不存在");
        }
        return dv;
    }

    public Map<String, String> diff(String docId, int v1, int v2) {
        DocumentVersion ver1 = getVersion(docId, v1);
        DocumentVersion ver2 = getVersion(docId, v2);
        return Map.of(
                "v1", ver1.getContentHtml() != null ? ver1.getContentHtml() : "",
                "v2", ver2.getContentHtml() != null ? ver2.getContentHtml() : "",
                "v1Version", String.valueOf(v1),
                "v2Version", String.valueOf(v2)
        );
    }

    public PageResult<Document> listBySpace(String spaceId, String folderId, int page, int size,
                                             String keyword, String status) {
        LambdaQueryWrapper<Document> qw = new LambdaQueryWrapper<>();
        qw.eq(Document::getSpaceId, spaceId);
        if (folderId != null) {
            qw.eq(Document::getFolderId, folderId);
        }
        if (keyword != null && !keyword.isBlank()) {
            qw.like(Document::getTitle, keyword);
        }
        if (status != null && !status.isBlank()) {
            qw.eq(Document::getStatus, status);
        }
        qw.orderByDesc(Document::getUpdatedAt);
        Page<Document> p = documentMapper.selectPage(new Page<>(page, size), qw);
        return new PageResult<>(p.getRecords(), p.getTotal(), page, size);
    }

    public List<Document> getRecent(int limit) {
        int safeLimit = Math.min(limit, 50);
        Page<Document> page = new Page<>(1, safeLimit);
        LambdaQueryWrapper<Document> qw = new LambdaQueryWrapper<>();
        qw.orderByDesc(Document::getUpdatedAt);
        return documentMapper.selectPage(page, qw).getRecords();
    }

    public Document restoreVersion(String docId, Integer versionNumber, String userId) {
        Document current = documentMapper.selectById(docId);
        if (current == null) {
            throw new BizException(404, "文档不存在");
        }

        DocumentVersion target = documentVersionMapper.selectOne(
                new LambdaQueryWrapper<DocumentVersion>()
                        .eq(DocumentVersion::getDocumentId, docId)
                        .eq(DocumentVersion::getVersionNumber, versionNumber));
        if (target == null) {
            throw new BizException(404, "要恢复的历史版本不存在");
        }

        // 1. 将当前文档状态备份为新版本记录
        DocumentVersion dv = new DocumentVersion();
        dv.setId(IdUtil.fastSimpleUUID());
        dv.setDocumentId(docId);
        dv.setVersionNumber(current.getVersion());
        dv.setContentJson(current.getContentJson());
        dv.setContentHtml(current.getContentHtml());
        dv.setChangeSummary("回滚至版本 V" + versionNumber);
        dv.setCreatedBy(userId);
        documentVersionMapper.insert(dv);

        // 2. 恢复至目标版本内容，版本号递增
        current.setContentJson(target.getContentJson());
        current.setContentHtml(target.getContentHtml());
        current.setVersion(current.getVersion() + 1);
        current.setUpdatedBy(userId);
        documentMapper.updateById(current);

        // 3. 异步更新 ES 索引
        rabbitTemplate.convertAndSend("doc.change", "doc.updated", docId);

        return documentMapper.selectById(docId);
    }
}
